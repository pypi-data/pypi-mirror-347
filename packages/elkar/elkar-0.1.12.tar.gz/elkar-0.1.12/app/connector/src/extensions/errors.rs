use axum::{
    body::Body,
    response::{IntoResponse, Response},
    Json,
};

use diesel::{result::Error as DieselError, ConnectionError};
use handlebars::RenderError;
use http::StatusCode;
use serde::Serialize;
use std::fmt::{Debug, Display};
use std::io::Error as IoError;
use std::{error::Error, sync::LockResult};
use supabase::SupabaseError;
use tokio::task::JoinError;

#[derive(Debug)]
pub struct ServiceError {
    status_code: StatusCode,
    error_type: String,
    details: Option<String>,
}

impl Display for ServiceError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} ({})",
            self.error_type,
            self.details.as_ref().unwrap_or(&"".to_string())
        )
    }
}

#[derive(Serialize, Debug)]
pub struct ServiceErrorResponse {
    error_type: String,
    details: Option<String>,
}

impl From<ServiceError> for ServiceErrorResponse {
    fn from(service_error: ServiceError) -> Self {
        Self {
            error_type: service_error.error_type,
            details: service_error.details,
        }
    }
}

impl From<&ServiceError> for ServiceErrorResponse {
    fn from(service_error: &ServiceError) -> Self {
        Self {
            error_type: service_error.error_type.clone(),
            details: service_error.details.clone(),
        }
    }
}

pub type AppResult<T> = Result<T, BoxedAppError>;

impl Default for ServiceError {
    fn default() -> Self {
        Self::new()
    }
}

impl ServiceError {
    pub fn new() -> Self {
        Self {
            status_code: StatusCode::INTERNAL_SERVER_ERROR,
            error_type: "Internal Server Error".to_string(),
            details: None,
        }
    }

    pub fn status_code(mut self, status_code: StatusCode) -> Self {
        self.status_code = status_code;
        self
    }
    pub fn error_type(mut self, error_type: impl ToString) -> Self {
        self.error_type = error_type.to_string();
        self
    }
    pub fn details(mut self, details: impl ToString) -> Self {
        self.details = Some(details.to_string());
        self
    }
}

impl IntoResponse for ServiceError {
    fn into_response(self) -> Response<Body> {
        (self.status_code, Json(ServiceErrorResponse::from(self))).into_response()
    }
}
pub type BoxedAppError = Box<dyn AppError>;

pub trait AppError: Debug + Display + Send + 'static {
    fn response(&self) -> axum::response::Response;

    fn cause(&self) -> Option<&dyn AppError> {
        None
    }
}

impl IntoResponse for BoxedAppError {
    fn into_response(self) -> Response<Body> {
        self.response()
    }
}

impl AppError for BoxedAppError {
    fn response(&self) -> Response {
        (**self).response()
    }
    fn cause(&self) -> Option<&dyn AppError> {
        (**self).cause()
    }
}

fn server_error_response(error: String) -> axum::response::Response {
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Json(ServiceErrorResponse {
            error_type: "Internal Server Error".to_string(),
            details: Some(error),
        }),
    )
        .into_response()
}

impl AppError for ServiceError {
    fn response(&self) -> axum::response::Response {
        let error = self.error_type.clone();
        let details = self.details.clone().unwrap_or("".to_string());
        let error_format = format!("{}: {}", error, details);
        tracing::error!(error = %error_format );
        sentry::capture_message(&error_format, sentry::Level::Error);

        (self.status_code, Json(ServiceErrorResponse::from(self))).into_response()
    }
}

impl<E: Error + Send + 'static> AppError for E {
    fn response(&self) -> axum::response::Response {
        tracing::error!(error = %self, "Internal Server Error");
        let error_format = format!("{:?}", self);
        sentry::configure_scope(|scope| {
            scope.set_fingerprint(Some([error_format.as_str()].as_ref()));
        });
        sentry::capture_error(self);

        server_error_response(self.to_string())
    }
}

impl From<ServiceError> for BoxedAppError {
    fn from(error: ServiceError) -> Self {
        Box::new(error)
    }
}

impl From<DieselError> for BoxedAppError {
    fn from(error: DieselError) -> Self {
        Box::new(
            ServiceError::new()
                .status_code(StatusCode::INTERNAL_SERVER_ERROR)
                .error_type("Database Error".to_string())
                .details(error.to_string()),
        )
    }
}

impl From<RenderError> for BoxedAppError {
    fn from(error: RenderError) -> Self {
        Box::new(
            ServiceError::new()
                .status_code(StatusCode::INTERNAL_SERVER_ERROR)
                .error_type("Template Error".to_string())
                .details(error.to_string()),
        )
    }
}

impl From<anyhow::Error> for BoxedAppError {
    fn from(error: anyhow::Error) -> Self {
        Box::new(
            ServiceError::new()
                .status_code(StatusCode::INTERNAL_SERVER_ERROR)
                .error_type("Internal Service Error".to_string())
                .details(error.to_string()),
        )
    }
}

impl From<serde_json::Error> for BoxedAppError {
    fn from(error: serde_json::Error) -> Self {
        Box::new(
            ServiceError::new()
                .status_code(StatusCode::INTERNAL_SERVER_ERROR)
                .error_type("Serde JSON Error".to_string())
                .details(error.to_string()),
        )
    }
}

impl From<JoinError> for BoxedAppError {
    fn from(error: JoinError) -> Self {
        Box::new(
            ServiceError::new()
                .status_code(StatusCode::INTERNAL_SERVER_ERROR)
                .error_type("Async Tokio Error".to_string())
                .details(error.to_string()),
        )
    }
}

impl<G> From<LockResult<G>> for BoxedAppError {
    fn from(_error: LockResult<G>) -> Self {
        Box::new(
            ServiceError::new()
                .status_code(StatusCode::INTERNAL_SERVER_ERROR)
                .error_type("mutex lock error".to_string()),
        )
    }
}

impl From<SupabaseError> for BoxedAppError {
    fn from(error: SupabaseError) -> Self {
        Box::new(
            ServiceError::new()
                .status_code(StatusCode::INTERNAL_SERVER_ERROR)
                .error_type("Supabase Error".to_string())
                .details(format!("{:?}", error)),
        )
    }
}

impl From<IoError> for BoxedAppError {
    fn from(error: IoError) -> Self {
        Box::new(
            ServiceError::new()
                .status_code(StatusCode::INTERNAL_SERVER_ERROR)
                .error_type("IO Error".to_string())
                .details(format!("{:?}", error)),
        )
    }
}

impl From<ConnectionError> for BoxedAppError {
    fn from(error: ConnectionError) -> Self {
        Box::new(
            ServiceError::new()
                .status_code(StatusCode::INTERNAL_SERVER_ERROR)
                .error_type("Diesel Connection Error".to_string())
                .details(format!("{:?}", error)),
        )
    }
}
