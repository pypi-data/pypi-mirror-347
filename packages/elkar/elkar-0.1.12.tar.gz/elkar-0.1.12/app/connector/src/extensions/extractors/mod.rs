pub mod api_key_context;
pub mod query_params;
pub mod user_context;
