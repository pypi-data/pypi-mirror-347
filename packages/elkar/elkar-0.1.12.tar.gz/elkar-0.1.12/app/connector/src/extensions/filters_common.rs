use std::{fmt::Display, str::FromStr};

use serde::*;

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub enum FilterType {
    AnyOf,
    AllOf,
    NoneOf,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, Default)]
#[serde(rename_all = "camelCase")]
pub enum ComparisonOperator {
    Eq,
    Ne,
    #[default]
    Gt,
    Gte,
    Lt,
    Lte,
}

impl Display for ComparisonOperator {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Eq => write!(f, "eq"),
            Self::Ne => write!(f, "ne"),
            Self::Gt => write!(f, "gt"),
            Self::Gte => write!(f, "gte"),
            Self::Lt => write!(f, "lt"),
            Self::Lte => write!(f, "lte"),
        }
    }
}
impl FromStr for ComparisonOperator {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "eq" => Ok(Self::Eq),
            "ne" => Ok(Self::Ne),
            "gt" => Ok(Self::Gt),
            "gte" => Ok(Self::Gte),
            "lt" => Ok(Self::Lt),
            "lte" => Ok(Self::Lte),
            _ => Err(anyhow::anyhow!("Invalid comparison operator")),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CounterFilter<T, V> {
    pub filter: T,
    pub operator: ComparisonOperator,
    pub value: V,
}

impl<T: Display, V: Display> Display for CounterFilter<T, V> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}:{}", self.filter, self.operator, self.value)
    }
}

impl<T: FromStr, V: Default + FromStr> FromStr for CounterFilter<T, V> {
    type Err = anyhow::Error;

    fn from_str(filter: &str) -> Result<Self, Self::Err> {
        let mut parts = filter.split(":");
        let filter = parts
            .next()
            .ok_or(anyhow::anyhow!(
                "Invalid filter for type {}",
                std::any::type_name::<T>()
            ))?
            .parse::<T>()
            .map_err(|_| {
                anyhow::anyhow!("Invalid filter for type {}", std::any::type_name::<T>())
            })?;

        let operator = parts
            .next()
            .unwrap_or_default()
            .parse::<ComparisonOperator>()
            .map_err(|_| anyhow::anyhow!("Invalid filter operator",))?;
        let value = parts
            .next()
            .unwrap_or_default()
            .parse::<V>()
            .map_err(|_| anyhow::anyhow!("Invalid filter value"))?;
        Ok(Self {
            filter,
            operator,
            value,
        })
    }
}

impl<T, V: Into<sea_query::Value> + Clone> CounterFilter<T, V> {
    pub fn apply_operator(&self, query: &mut sea_query::SelectStatement, col: sea_query::Expr) {
        match self.operator {
            ComparisonOperator::Eq => {
                query.and_where(col.eq(self.value.clone()));
            }
            ComparisonOperator::Ne => {
                query.and_where(col.ne(self.value.clone()));
            }
            ComparisonOperator::Gt => {
                query.and_where(col.gt(self.value.clone()));
            }
            ComparisonOperator::Gte => {
                query.and_where(col.gte(self.value.clone()));
            }
            ComparisonOperator::Lt => {
                query.and_where(col.lt(self.value.clone()));
            }
            ComparisonOperator::Lte => {
                query.and_where(col.lte(self.value.clone()));
            }
        }
    }
}
