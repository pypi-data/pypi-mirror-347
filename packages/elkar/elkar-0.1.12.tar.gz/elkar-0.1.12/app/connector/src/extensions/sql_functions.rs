use diesel::{
    define_sql_function,
    sql_types::{Nullable, SingleValue, Text, Timestamp},
};
use sea_query::{Iden, Write};

define_sql_function!(fn date_trunc(interval: Text, timestamp: Timestamp) -> Timestamp);

define_sql_function!(fn date(timestamp: Timestamp) -> Date);

define_sql_function! {
    #[aggregate]
    #[sql_name = "ARRAY_AGG"]
    fn array_agg<T:SingleValue>(expr: T) -> Array<T>;
}
define_sql_function! {fn coalesce<T:SingleValue>(x:Nullable<T>,y:T) -> T;}

// SeaQuery
pub struct ArrayAgg;

impl Iden for ArrayAgg {
    fn unquoted(&self, s: &mut dyn Write) {
        write!(s, "ARRAY_AGG").unwrap();
    }
}

pub struct Format;

impl Iden for Format {
    fn unquoted(&self, s: &mut dyn Write) {
        write!(s, "FORMAT").unwrap();
    }
}

pub struct JsonBObject;

impl Iden for JsonBObject {
    fn unquoted(&self, s: &mut dyn Write) {
        write!(s, "JSONB_OBJECT").unwrap();
    }
}

pub struct BoolOr;

impl Iden for BoolOr {
    fn unquoted(&self, s: &mut dyn Write) {
        write!(s, "BOOL_OR").unwrap();
    }
}

pub struct AnyValue;

impl Iden for AnyValue {
    fn unquoted(&self, s: &mut dyn Write) {
        write!(s, "ANY_VALUE").unwrap();
    }
}
