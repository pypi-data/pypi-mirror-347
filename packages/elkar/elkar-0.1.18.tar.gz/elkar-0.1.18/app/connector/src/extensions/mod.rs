pub mod async_database;
pub mod errors;
pub mod extractors;
pub mod filters_common;
pub mod helpers;
pub mod pagination;
pub mod query_common;
pub mod sentry;
pub mod sql_functions;
pub mod token;
use secrecy::SecretString;

#[derive(Debug, Clone)]
pub struct SupabaseConfig {
    pub url: SecretString,
    pub secret_key: SecretString,
    pub admin_key: SecretString,
}

#[derive(Debug, Clone)]
pub struct DatabaseConfig {
    pub app_user_url: SecretString,
    pub admin_url: SecretString,
    pub no_rls_user_url: SecretString,
}
#[derive(Debug, Clone)]
pub struct NotionConfig {
    pub client_id: SecretString,
    pub client_secret: SecretString,
    pub redirect_uri: SecretString,
}

#[derive(Debug, Clone)]
pub struct AppConfig {
    pub environment: Environment,
    pub frontend_url: String,
    pub sentry_dsn: SecretString,
    pub supabase: SupabaseConfig,
    pub database: DatabaseConfig,
}

lazy_static! {
    pub static ref APP_CONFIG: AppConfig = AppConfig {
        environment: Environment::try_from(dotenv::var("ENVIRONMENT").unwrap_or_default()).unwrap(),
        frontend_url: dotenv::var("FRONT_URL").unwrap(),
        sentry_dsn: dotenv::var("SENTRY_DSN").unwrap().into(),
        supabase: SupabaseConfig {
            url: dotenv::var("SUPABASE_URL").unwrap().into(),
            secret_key: dotenv::var("SUPABASE_SECRET_KEY").unwrap().into(),
            admin_key: dotenv::var("SUPABASE_ADMIN_KEY").unwrap().into(),
        },
        database: DatabaseConfig {
            app_user_url: dotenv::var("APP_USER_DATABASE_URL").unwrap().into(),
            admin_url: dotenv::var("DATABASE_URL").unwrap().into(),
            no_rls_user_url: dotenv::var("NO_RLS_USER_DATABASE_URL").unwrap().into(),
        },
    };
}

/// The possible runtime environment for our application.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Environment {
    Local,
    Development,
    Production,
}
impl Environment {
    pub fn as_str(&self) -> &'static str {
        match self {
            Environment::Local => "local",
            Environment::Development => "development",
            Environment::Production => "production",
        }
    }
}

impl TryFrom<String> for Environment {
    type Error = String;
    fn try_from(s: String) -> Result<Self, Self::Error> {
        match s.to_lowercase().as_str() {
            "local" => Ok(Self::Local),
            "development" => Ok(Self::Development),
            "production" => Ok(Self::Production),
            other => Err(format!(
                "{} is not a supported environment. Use either `local`, `development` or `production`.",
                other
            )),
        }
    }
}
