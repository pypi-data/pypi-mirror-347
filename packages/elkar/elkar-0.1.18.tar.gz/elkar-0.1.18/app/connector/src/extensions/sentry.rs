use sentry::integrations::tracing::EventFilter;

use tracing::Metadata;
use tracing::{Level, level_filters::LevelFilter};
use tracing_subscriber::filter::EnvFilter;
use tracing_subscriber::prelude::*;

pub fn init() {
    init_with_default_level(LevelFilter::INFO)
}

pub fn event_filter(metadata: &Metadata<'_>) -> EventFilter {
    match metadata.level() {
        &Level::ERROR => EventFilter::Exception,
        &Level::WARN | &Level::INFO => EventFilter::Breadcrumb,
        &Level::DEBUG | &Level::TRACE => EventFilter::Ignore,
    }
}

fn init_with_default_level(level: LevelFilter) {
    let env_filter = EnvFilter::builder()
        .with_default_directive(level.into())
        .from_env_lossy();

    let log_layer = tracing_subscriber::fmt::layer()
        .pretty()
        .with_file(true)
        .with_line_number(true)
        .without_time()
        .with_ansi(false)
        .with_filter(env_filter);

    let sentry_layer = sentry::integrations::tracing::layer()
        .event_filter(event_filter)
        .with_filter(LevelFilter::INFO);

    tracing_subscriber::registry()
        .with(log_layer)
        .with(sentry_layer)
        .init();
}
