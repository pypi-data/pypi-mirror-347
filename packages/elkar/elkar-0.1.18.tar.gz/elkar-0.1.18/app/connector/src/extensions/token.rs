use std::fmt::Display;

use http::HeaderMap;
use jsonwebtoken::{Algorithm, DecodingKey, Validation, decode};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::{APP_CONFIG, Environment};

#[derive(Debug)]
pub struct SupabaseToken<'a>(&'a str);

impl Display for SupabaseToken<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Supabase Token Error: {}", self.0)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum AudEnum {
    String(String),
    Vec(Vec<String>),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Amr {
    pub method: String,
    pub timestamp: i64,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DecodedSupabaseToken {
    pub aud: AudEnum,
    pub exp: i64,
    pub iat: i64,
    pub iss: String,
    pub sub: Uuid,
    // pub encrypted_password: Option<String>,
    pub email: Option<String>,
    pub phone: Option<String>,
    pub role: Option<String>,
    pub aal: Option<String>,
    pub session_id: Option<String>,
    pub amr: Option<Vec<Amr>>,
}

impl<'a> SupabaseToken<'a> {
    pub fn new(token: &'a str) -> Self {
        Self(token)
    }
    pub fn decode(&self) -> Result<DecodedSupabaseToken, jsonwebtoken::errors::Error> {
        let secret_key = dotenv::var("SUPABASE_SECRET_KEY").unwrap();
        let mut validation = Validation::new(Algorithm::HS256);
        validation.set_audience(&["authenticated"]);
        validation.set_required_spec_claims(&["amr", "aal"]);
        validation.validate_exp = true;
        let decoded_token = decode::<DecodedSupabaseToken>(
            self.0,
            &DecodingKey::from_secret(secret_key.as_ref()),
            &validation,
        )?;

        // if APP_CONFIG.environment == Environment::Production {
        //     let amr = decoded_token.claims.amr.clone().unwrap_or_default();
        //     if amr.iter().any(|a| a.method == "password")
        //         && decoded_token.claims.aal != Some("aal2".to_string())
        //     {
        //         return Err(jsonwebtoken::errors::ErrorKind::MissingRequiredClaim(
        //             "Missing 2FA".to_string(),
        //         )
        //         .into());
        //     }
        // }
        Ok(decoded_token.claims)
    }
}

#[derive(Debug, Clone, Copy, Default)]
pub struct TokenError<'a>(&'a str);

impl Display for TokenError<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Token Error: {}", self.0)
    }
}

pub fn extract_token(header_map: &HeaderMap) -> Result<String, TokenError> {
    let bearer_token = header_map
        .get("Authorization")
        .ok_or(TokenError("Missing Authorization Header"))?
        .to_str()
        .map_err(|_| TokenError("Failed to parse Token"))?;
    let token = bearer_token
        .strip_prefix("Bearer ")
        .ok_or(TokenError("Failed to parse Token"))?;
    Ok(token.to_string())
}
