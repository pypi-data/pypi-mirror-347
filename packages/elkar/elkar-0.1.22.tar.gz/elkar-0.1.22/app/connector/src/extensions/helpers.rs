use std::{future::Future, time::Duration};
use tokio::task::{AbortHandle, JoinError, JoinSet};

pub struct BoundedJoinSet<T> {
    join_set: JoinSet<T>,
    max_concurrency: usize,
    collection: Vec<T>,
    errors: Vec<JoinError>,
    sleep_duration: Option<Duration>,
}

impl<T> BoundedJoinSet<T> {
    pub fn new(max_concurrency: usize) -> Self {
        Self {
            join_set: JoinSet::new(),
            max_concurrency,
            collection: Vec::new(),
            errors: Vec::new(),
            sleep_duration: None,
        }
    }

    pub fn set_sleep_duration(&mut self, sleep_duration: Option<Duration>) {
        self.sleep_duration = sleep_duration;
    }

    pub fn into_collection(self) -> Vec<T> {
        self.collection
    }

    pub fn collection(&self) -> &Vec<T> {
        &self.collection
    }

    pub fn into_errors(self) -> Vec<JoinError> {
        self.errors
    }

    pub fn errors(&self) -> &Vec<JoinError> {
        &self.errors
    }

    pub fn has_errors(&self) -> bool {
        !self.errors.is_empty()
    }

    pub fn trace_errors(&self) {
        for error in &self.errors {
            tracing::error!("BoundedJoinSet Error: {:?}", error);
        }
    }
}
impl<T: 'static> BoundedJoinSet<T> {
    pub async fn spawn<F>(&mut self, future: F) -> AbortHandle
    where
        F: Future<Output = T>,
        F: Send + 'static,
        T: Send,
    {
        while self.join_set.len() >= self.max_concurrency {
            if let Some(result) = self.join_set.join_next().await {
                match result {
                    Ok(item) => self.collection.push(item),
                    Err(e) => self.errors.push(e),
                }
            }
            if let Some(sleep_duration) = self.sleep_duration {
                tokio::time::sleep(sleep_duration).await;
            }
        }
        self.join_set.spawn(future)
    }

    pub async fn join_next(&mut self) -> Option<Result<T, JoinError>> {
        self.join_set.join_next().await
    }

    pub async fn join_all(&mut self) {
        while let Some(result) = self.join_set.join_next().await {
            match result {
                Ok(item) => self.collection.push(item),
                Err(e) => self.errors.push(e),
            }
        }
    }
}
