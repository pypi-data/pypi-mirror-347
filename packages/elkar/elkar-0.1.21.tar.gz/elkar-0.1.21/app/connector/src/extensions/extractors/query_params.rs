use std::fmt::Debug;

use axum::{
    extract::FromRequestParts,
    http::{request::Parts, StatusCode},
};

use serde_querystring::ParseMode;

use crate::extensions::errors::{BoxedAppError, ServiceError};

#[derive(Debug, Clone, Copy, Default)]
pub struct Qs<T>(pub T);

impl<S, T> FromRequestParts<S> for Qs<T>
where
    T: serde::de::DeserializeOwned + Debug + Send,
    S: Send + Sync,
{
    type Rejection = BoxedAppError;

    async fn from_request_parts(parts: &mut Parts, _state: &S) -> Result<Self, Self::Rejection> {
        let query = parts.uri.query().unwrap_or("");
        let qs_value = serde_querystring::from_str(query, ParseMode::Duplicate).map_err(|e| {
            ServiceError::new()
                .status_code(StatusCode::BAD_REQUEST)
                .error_type("Invalid query string".to_string())
                .details(e.to_string())
        })?;
        Ok(Qs(qs_value))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use chrono::{DateTime, Utc};
    use insta::assert_json_snapshot;
    use serde::{Deserialize, Serialize};

    // Add chrono with serde feature for DateTime serialization

    #[derive(Debug, Serialize, Deserialize)]
    struct Query {
        per_page: u32,
        page: u32,
        subject_search: String,
        date_gte: DateTime<Utc>,
    }

    #[tokio::test]
    async fn test_qs() {
        let query = "per_page=15&page=1&subject_search=&date_gte=2024-10-08T22%3A00%3A00.000Z";
        let qs_value: Query = serde_querystring::from_str(query, ParseMode::Duplicate)
            .map_err(|e| {
                ServiceError::new()
                    .status_code(StatusCode::BAD_REQUEST)
                    .error_type("Invalid query string".to_string())
                    .details(e.to_string())
            })
            .unwrap();

        assert_json_snapshot!(qs_value)
    }
}
