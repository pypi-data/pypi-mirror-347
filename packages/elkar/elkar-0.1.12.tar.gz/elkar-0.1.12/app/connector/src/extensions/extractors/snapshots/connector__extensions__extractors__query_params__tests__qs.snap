---
source: src/extensions/extractors/query_params.rs
expression: qs_value
---
{
  "per_page": 15,
  "page": 1,
  "subject_search": "",
  "date_gte": "2024-10-08T22:00:00Z"
}
